### PR TITLE
Implement r_str_cat and other random cleanups

### PR DESCRIPTION
**Pull request by trufae**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

Leftovers from the hint-glitch pr

### CODE REVIEW

**Comment by XVilka** [comment]
Either there is somewhere an off-by-one mistake or please fix the tests.

File: libr/core/disasm.c [hunk 1]
```diff
@@ -1613,7 +1619,7 @@ static void ds_pre_xrefs(RDisasmState *ds, bool no_fcnlines) {
 	ds->line_col = ds->prev_line_col;
 	ds_print_lines_left (ds);
 	if (!ds->show_offset && ds->show_marks) {
-		ds_print_show_cursor (ds);
+		ds_print_show_cursor (ds, 0);
```
Comment (ret2libc) [inline libr/core/disasm.c:1622]:
```suggestion
		ds_print_show_cursor (ds, false);
```

File: libr/core/disasm.c [hunk 2]
```diff
@@ -3614,16 +3618,18 @@ static bool ds_print_core_vmode(RDisasmState *ds, int pos) {
 		if (ds_print_core_vmode_jump_hit (ds, pos)) {
 			gotShortcut = true;
 		}
+		if (ds->asm_hint_pos <= 0) {
+			gotShortcut = false;
+			// ds_print_show_cursor (ds, false);
```
Comment (ret2libc) [inline libr/core/disasm.c:3623]:
```suggestion
```

**Comment by trufae** [comment]
cursor and asm hints only appear in visual mode, so not sure if i can add tests for this..

**Comment by trufae** [comment]
Please review

**Review by thestr4ng3r — APPROVED** [review]

File: libr/core/disasm.c [hunk 3]
```diff
@@ -1598,6 +1601,9 @@ static void ds_print_show_cursor(RDisasmState *ds) {
 			res[i] = '0' + (diff % 10);
 		}
 	}
+	if (has_hint) {
+		strcpy (res, res + 3);
```
Comment (ret2libc) [inline libr/core/disasm.c:1605]: I think this involves moving data between overlapping strings, right? If so, man page says that `strcpy`' strings may not overlap.

**Review by ret2libc — CHANGES_REQUESTED** [review]

File: libr/core/disasm.c (hunk 3 above)
Comment (trufae) [reply libr/core/disasm.c:1605]: Indeed. Should be using memmove. Do u think we should provide r_str_cpy as a safe alternative?

Comment (ret2libc) [reply libr/core/disasm.c:1605]: There is already:
```C
#define r_str_cpy(x,y) memmove(x,y,strlen(y)+1);
```

Comment (trufae) [reply libr/core/disasm.c:1605]: i see some missing parenthesis in this macro. ill fix that too

File: libr/core/disasm.c (hunk 1 above)
Comment (ret2libc) [reply libr/core/disasm.c:1629]: See ^ please

File: libr/core/disasm.c [hunk 4]
```diff
@@ -5483,11 +5488,12 @@ R_API int r_core_print_disasm(RPrint *p, RCore *core, ut64 addr, ut8 *buf, int l
 		////
 		int mi_type;
 		bool mi_found = ds_print_meta_infos (ds, buf, len, idx, &mi_type);
+		bool hasHint = false;
```
Comment (ret2libc) [inline libr/core/disasm.c:5491]: not a blocker, but `s/hasHint/has_hint`.

Comment (trufae) [reply libr/core/disasm.c]: fixed

File: libr/core/disasm.c (hunk 1 above)
Comment (trufae) [reply libr/core/disasm.c:1629]: I fixed this already but the commit got lost becausr when gh updates a change it breaks the whole commit history and i purge it and chcerrypick one by one when pulling

Comment (ret2libc) [reply libr/core/disasm.c:1629]: Please re-commit it. GH does not break the history, most likely you rebased before pulling and getting the newer commit commited directly by GH.

Comment (trufae) [reply libr/core/disasm.c]: done

**Review by thestr4ng3r — APPROVED** [review]

**Comment by ret2libc** [comment]
Actually... do you have a precise reproducer for the issue you mentioned in the description? I cannot reproduce and for me `asm.hint.pos=0` and `asm.hint.pos=-1` already work well.

**Comment by trufae** [comment]
As i said @ret2libc this issue hyappens ONLY in visual mode. so it cant be tested

**Comment by ret2libc** [comment]
> As i said @ret2libc this issue hyappens ONLY in visual mode. so it cant be tested

I am testing visual mode indeed. Manually. And I can't reproduce. Could you provide a reproducer?

**Comment by trufae** [comment]
Try with arm1.bin

**Review by thestr4ng3r — APPROVED** [review]

**Pull request merged**